drivers/sensor/qdec_stm32/qdec_stm32.c

```diff
@@ -0,0 +1,151 @@
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+#include <stm32_ll_tim.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(qdec_stm32, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Device constant configuration parameters */
+struct qdec_stm32_dev_cfg {
+	const struct pinctrl_dev_config *pin_config;
+	struct stm32_pclken pclken;
+	TIM_TypeDef *timer_inst;
+	uint16_t prescaler;
+	uint8_t is_input_polarity_inverted;
+	uint8_t input_filtering_level;
+};
+
+/* Device run time data */
+struct qdec_stm32_dev_data {
+	int32_t position;
+};
+
+static int qdec_stm32_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct qdec_stm32_dev_data *dev_data = dev->data;
+	const struct qdec_stm32_dev_cfg *dev_cfg = dev->config;
+
+	if ((chan != SENSOR_CHAN_ALL) && (chan != SENSOR_CHAN_ROTATION)) {
+		return -ENOTSUP;
+	}
+
+	dev_data->position = (int32_t) LL_TIM_GetCounter(dev_cfg->timer_inst);
+
+	return 0;
+}
+
+static int qdec_stm32_get(const struct device *dev, enum sensor_channel chan,
+			struct sensor_value *val)
+{
+	struct qdec_stm32_dev_data *const dev_data = dev->data;
+
+	if (chan == SENSOR_CHAN_ROTATION) {
+		val->val1 = dev_data->position;
+		val->val2 = 0;
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int qdec_stm32_initialize(const struct device *dev)
+{
+	__ASSERT_NO_MSG(dev != NULL);
+	const struct qdec_stm32_dev_cfg *const dev_cfg = dev->config;
+	int retval;
+	LL_TIM_ENCODER_InitTypeDef init_props;
+
+	retval = pinctrl_apply_state(dev_cfg->pin_config, PINCTRL_STATE_DEFAULT);
+	if (retval < 0) {
+		return retval;
+	}
+
+	if (!device_is_ready(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE))) {
+		LOG_ERR("Clock control device not ready");
+		return -ENODEV;
+	}
+
+	retval = clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
+			     (clock_control_subsys_t)&dev_cfg->pclken);
+	if (retval < 0) {
+		LOG_ERR("Could not initialize clock");
+		return retval;
+	}
+
+	LL_TIM_ENCODER_StructInit(&init_props);
+
+	if (dev_cfg->is_input_polarity_inverted) {
+		init_props.IC1ActiveInput = LL_TIM_IC_POLARITY_FALLING;
+		init_props.IC2ActiveInput = LL_TIM_IC_POLARITY_FALLING;
+	}
+
+	init_props.IC1Filter = dev_cfg->input_filtering_level * LL_TIM_IC_FILTER_FDIV1_N2;
+	init_props.IC2Filter = dev_cfg->input_filtering_level * LL_TIM_IC_FILTER_FDIV1_N2;
+
+	if (IS_TIM_32B_COUNTER_INSTANCE(dev_cfg->timer_inst)) {
+		LL_TIM_SetAutoReload(dev_cfg->timer_inst, 0xFFFFFFFF);
+	} else {
+		LL_TIM_SetAutoReload(dev_cfg->timer_inst, 0xFFFF);
+	}
+
+	if (LL_TIM_ENCODER_Init(dev_cfg->timer_inst, &init_props) != SUCCESS) {
+		LOG_ERR("Initalization failed");
+		return -EIO;
+	}
+
+	LL_TIM_SetPrescaler(dev_cfg->timer_inst, dev_cfg->prescaler);
+
+	LL_TIM_EnableCounter(dev_cfg->timer_inst);
+
+	LOG_INF("Device %s initialized", dev->name);
```

Comment:
`LOG_DBG`? is it actually worth?

Reply:
it was present in the driver I used as reference, so I kept it, but you're right. I can remove it